langchain/agents/agent_toolkits/openapi/planner.py

```diff
@@ -0,0 +1,375 @@
+
+Output:""",
+            input_variables=["response", "instructions"],
+        ),
+    )
+
+    def _run(self, text: str) -> str:
+        try:
+            data = json.loads(text)
+            response = self.requests_wrapper.post(data["url"], data["data"])
+            response = response[: self.response_length]
+            # print(f"Request tool input: {data}\nResponse: {response}")
+            return self.llm_chain.predict(
+                response=response, instructions=data["output_instructions"]
+            ).strip()
+        except Exception as e:
+            return repr(e)
+
+    async def _arun(self, text: str) -> str:
+        pass
+
+
+#
+# Orchestrator, planner, controller.
+#
+def _create_api_planner_tool(
+    api_spec: ReducedOpenAPISpec, llm: BaseLanguageModel = DEFAULT_LLM
+) -> Tool:
+    endpoint_descriptions = [
+        f"{name} {description}" for name, description, _ in api_spec.endpoints
+    ]
+    prompt = PromptTemplate(
+        template=API_PLANNER_PROMPT,
+        input_variables=["query"],
+        partial_variables={"endpoints": "- " + "- ".join(endpoint_descriptions)},
+    )
+    chain = LLMChain(llm=llm, prompt=prompt)
+    tool = Tool(
+        name=API_PLANNER_TOOL_NAME,
+        description=API_PLANNER_TOOL_DESCRIPTION,
+        func=chain.run,
+    )
+    return tool
+
+
+def _create_api_controller_agent(
+    api_url: str,
+    api_docs: str,
+    requests_wrapper: RequestsWrapper,
+    llm: BaseLanguageModel = DEFAULT_LLM,
+) -> AgentExecutor:
+    tools = [
+        RequestsGetToolWithSummary(
+            requests_wrapper=requests_wrapper, response_length=MAX_RESPONSE_LENGTH
+        ),
+        RequestsPostToolWithSummary(
+            requests_wrapper=requests_wrapper, response_length=MAX_RESPONSE_LENGTH
+        ),
+    ]
+    prompt = PromptTemplate(
+        template=API_CONTROLLER_PROMPT,
+        input_variables=["input", "agent_scratchpad"],
+        partial_variables={
+            "api_url": api_url,
+            "api_docs": api_docs,
+            "tool_names": ", ".join([tool.name for tool in tools]),
+            "tool_descriptions": "\n".join(
+                [f"{tool.name}: {tool.description}" for tool in tools]
+            ),
+        },
+    )
+    agent = ZeroShotAgent(
+        llm_chain=LLMChain(llm=llm, prompt=prompt),
+        allowed_tools=[tool.name for tool in tools],
+    )
+    return AgentExecutor.from_agent_and_tools(agent=agent, tools=tools, verbose=True)
+
+
+def _create_api_controller_tool(
+    api_spec: ReducedOpenAPISpec, requests_wrapper: RequestsWrapper
+) -> Tool:
+    """Expose controller as a tool. The tool is invoked with a plan from the planner, and dynamically
+    creates a controller agent with relevant documentation only to constrain the context.
+    """
+
+    base_url = api_spec.servers[0]["url"]  # TODO: do better.
+
+    def _create_and_run_api_controller_agent(plan_str: str) -> str:
```

Comment:
Don't know if it'll make a difference, but I was thinking it could've been nice to use a CSV format for the list of endpoints and include an ID column, so that the plan can reference endpoint IDs. (Empirically, maybe it's difficult to coerce it to output the endpoint ID together with an instruction of what to do.)

```markdown
ID | VERB | PATH     | WHAT TO DO
------------------------------------------------------------------
1   |   GET |   search |   search for a given artist
2 ...
```

Reply:
I think it's reasonable to think about structuring the plan a bit, but this hasn't been the bottleneck so far so I was sticking with the simplest, most naive structure 😀